sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -190,9 +191,9 @@ Connection getConnection() throws Exception {
         basicDataSource.setPassword(getPassword());
         dataSource = basicDataSource;
       }
-      return (getUsername() == null)
-          ? dataSource.getConnection()
-          : dataSource.getConnection(getUsername(), getPassword());
+      return (getDataSource() != null && getUsername() != null)
```

Comment:
This is entangled/redundant with the flow of control and data above and I've spent a couple of minutes staring at this condition and still don't quite understand what it intends to do: looks like "if data source is not specified explicitly, then we've already set username and password at lines 190-191, so do not supply them again"?

- Why is this important - will something break if we supply them again?
- This will be much easier to read if you merge it into the branches of the if() above, i.e. have one return statement in the if() branch and one in the else() branch.